### PR TITLE
Fix 404 routing

### PR DESCRIPTION
**Pull request by gkallenberg**
Update routes.jsx to use named routes instead of `:type`
Use location.pathname in Application.jsx instead of `:type` which is no longer defined in routes.jsx
Update selectData.js to pass `next()` instead of redirecting to the baseUrl or base404 endpoints
Add logic to remove the Hero for 404 pages

### CODE REVIEW

File: src/app/routes/routes.jsx [hunk 1]
```diff
@@ -1,25 +1,32 @@
...
-      <Route path={`${appConfig.baseUrl}404`} component={Error404Page} />
-      <Route path={`${appConfig.baseUrl}:type`} component={Main} />
-      <Route path={`${appConfig.baseUrl}:type/:time`} component={Main} />
-      <Redirect from="*" to={`${appConfig.baseUrl}404`} />
-    </Route>
+    <Router>
+      <Route path={`${config.baseUrl}`} component={App}>
+        <IndexRoute component={Main} />
+        <Route path={`${config.base404}`} status={404} component={Error404Page} />
+        <Route path={`${config.baseUrl}staff-picks`} component={Main} />
```
Comment (EdwinGuzman) [inline src/app/routes/routes.jsx]: This doesn't work if you remove lines 21-23 and make the `:time` param optional?

File: src/app/components/Application/Application.jsx [hunk 2]
```diff
@@ -38,13 +38,19 @@ class App extends React.Component {
 
   render() {
     let heroData;
-    const type = !_isEmpty(this.props.params) ? this.props.params.type : undefined;
+    let heroDOM;
+    // Use pathname index as type since routes.jsx specifies them directly.
+    let type = this.props.location.pathname.split('/')[4];
```
Comment (ktp242) [inline src/app/components/Application/Application.jsx]: Eslint complains 'location' is missing in props validation.

File: src/app/routes/routes.jsx [hunk 3]
```diff
@@ -1,25 +1,32 @@
 // React library
 import React from 'react';
 // Import Router
-import { IndexRoute, Route, Redirect } from 'react-router';
+import { IndexRoute, Router, Route, Redirect } from 'react-router';
```
Comment (ktp242) [inline src/app/routes/routes.jsx]: `Redirect` is never used. We can remove it.

File: src/server/routes/selectData.js [hunk 4]
```diff
@@ -1,10 +1,7 @@
 import monthData from './monthData';
 import annualData from './annualData';
-import appConfig from '../../../appConfig';
```
Comment (ktp242) [inline src/server/routes/selectData.js]: I think we will need `appConfig`. It is used at Line 18 and 21.

Comment (gkallenberg) [reply src/server/routes/selectData.js]: Yup -- will add back -- hasty merge.

File: src/app/components/Application/Application.jsx (hunk 2 above)
Comment (ktp242) [inline src/app/components/Application/Application.jsx]: Probably to have a check here to see if `this.props.location.pathname` is a string.

File: src/app/components/Application/Application.jsx [hunk 5]
```diff
@@ -38,13 +38,19 @@ class App extends React.Component {
...
+      type = 'staffPicks';
+    }
 
     // Check if the params type is included in valid data set
     if (type && _allKeys(config.heroData).includes(type)) {
-      heroData = config.heroData[this.props.params.type];
-    } else {
-      heroData = config.heroData.staffPicks;
+      heroData = config.heroData[type] ? config.heroData[type] : config.heroData.staffPicks;
+      heroDOM = <Hero heroData={heroData} />;
```
Comment (ktp242) [inline src/app/components/Application/Application.jsx]: As the development branch updated. This part of scripts will need be updated too. For example, <Hero> now takes two props.

**Comment by ktp242** [comment]
Some situations I noticed

1. if we have the URL like this
http://local.nypl.org:3001/books-music-movies/recommendations/best-books/staff-picks/2017-06-01/fjlnfhlf%20jhnh
it will redirect to
http://local.nypl.org:3001/books-music-movies/recommendations/best-books/staff-picks/2017-06-01#fjlnfhlf%20jhnh
and still render the list
2. if we have the URL like this
http://local.nypl.org:3001/books-music-movies/recommendations/best-books/ya/2014/igjflglfjlgfgjlfg
it will redirect to http://local.nypl.org:3001/books-music-movies/recommendations/best-books/ya/2014#igjflglfjlgfgjlfg
and break the application
3. if we have URL like this
http://local.nypl.org:3001/books-music-movies/recommendations/best-books/hdfkjghkfghkfhg
this will lead to 404 content but the URL still remains

File: src/server/routes/selectData.js (hunk 4 above)
Comment (gkallenberg) [reply src/server/routes/selectData.js]: Added back. See below.

**Comment by EdwinGuzman** [comment]
To follow up on Kang's last comment, you can fix that by adding a check in `componentDidMount` in `main.jsx` to make sure that `pick` exists in line 39/40. That way, if pick doesn't exist in the list, then redirect because the user probably mistyped the id.

**Comment by gkallenberg** [comment]
I've addressed issues #1 and #2 of @ktp242's comment. #3 I think is fine the way it is for now. It may need a more practiced hand to add the redirect to the 404 somewhere where it doesn't re-introduce the 404 redirect loop. Pages on the main site typically serve the 404 content with the URL remaining as-is. See: https://www.nypl.org/lakjlskje

File: server.js [hunk 6]
```diff
@@ -54,6 +54,16 @@ nyplApiClient();
 
 app.use('/', expressRoutes);
 
+// Handle trailing slash
+app.use((req, res, next) => {
+  const test = /\?[^]*\//.test(req.url);
```
Comment (EdwinGuzman) [inline server.js]: A better variable name would be more helpful.

**Pull request merged**